### PR TITLE
4 Minor features

### PR DESCRIPTION
**Pull request by quazi-irfan**
New features in this pull request,
1. Signature at waiver agreement page in now case insensitive
2. If the user have a certificate and if the certification is expired, then user can not check-in. But user without any certification can still check in.
3. Both check-in and account creation page now accepts 6 digit member ID.
4. Use can not check-in if his account is suspended.

### CODE REVIEW

File: app/wama.web/AppString.resx [hunk 1]
```diff
@@ -120,6 +120,9 @@
   <data name="AccountPendingApprovalMessage" xml:space="preserve">
     <value>Account approval pending. Please contact front desk.</value>
   </data>
+  <data name="AccountSuspended" xml:space="preserve">
+    <value>Your account has been suspecded! Please contact the front desk.</value>
```
Comment (sakurano92) [inline app/wama.web/AppString.resx]: spelling mistake suspecded -> suspended please update

**Review by AboodV2 — COMMENTED** [review]
It looks good for me

File: app/wama.web/Controllers/CheckInController.cs [hunk 2]
```diff
@@ -44,6 +47,14 @@ public IActionResult Index()
                 {
                     SetErrorMessages(AppString.AccountPendingApprovalMessage);
                 }
+                else if (userAccount.IsSuspended)
+                {
+                    SetErrorMessages(AppString.AccountSuspended);
+                }
+                else if (certificateInfo != null && System.DateTimeOffset.Compare(System.DateTimeOffset.Now, certificateInfo.ExpiresOn) > 0)
+                {
+                    SetErrorMessages(AppString.CertificateExpired);
```
Comment (dubeme) [inline app/wama.web/Controllers/CheckInController.cs]: When a certification expires, a patron can still login.
Hence, you should figure out a means to pass a warning message to the Successful page.

File: app/wama.web/Controllers/CheckInController.cs [hunk 3]
```diff
@@ -44,6 +47,14 @@ public IActionResult Index()
                 {
                     SetErrorMessages(AppString.AccountPendingApprovalMessage);
                 }
+                else if (userAccount.IsSuspended)
+                {
+                    SetErrorMessages(AppString.AccountSuspended);
+                }
+                else if (certificateInfo != null && System.DateTimeOffset.Compare(System.DateTimeOffset.Now, certificateInfo.ExpiresOn) > 0)
```
Comment (dubeme) [inline app/wama.web/Controllers/CheckInController.cs]: The name certificationInfo can be simplified to certification. It's less redundant.

File: app/wama.web/Controllers/CheckInController.cs [hunk 4]
```diff
@@ -76,9 +87,9 @@ public IActionResult WaiverNotSigned()
         {
             var userAccount = await _UserAccountService.GetUserAccountAsync(memberId);
 
-            string userName = userAccount.FirstName + " " + userAccount.LastName;
+            string userName = userAccount.FirstName + " " + userAccount.LastName;
```
Comment (dubeme) [inline app/wama.web/Controllers/CheckInController.cs]: Don't just assume that the userAccount will always exist. 
Something weird can happen en-route to the DB, hence resulting in null being returned.

Perform proper checks for null on this variable.

File: app/wama.web/Views/CheckIn/Index.cshtml [hunk 5]
```diff
@@ -16,8 +16,8 @@
                        name="memberId"
                        id="check-in-user-id"
                        placeholder="Please enter your ID"
-                       pattern="[0-9]{7}"
-                       title="7 digit long numbers only!" />
+                       pattern="[0-9]{6,7}"
+                       title="6 or 7 digit long numbers only!" />
```
Comment (dubeme) [inline app/wama.web/Views/CheckIn/Index.cshtml:20]: Also, update the ViewModel used to generate these forms

**Review by dubeme — CHANGES_REQUESTED** [review]
Almost there, see comments

File: app/wama.web/Controllers/CheckInController.cs [hunk 6]
```diff
@@ -87,10 +87,15 @@ public IActionResult WaiverNotSigned()
         {
             var userAccount = await _UserAccountService.GetUserAccountAsync(memberId);
 
-            string userName = userAccount.FirstName + " " + userAccount.LastName;            
-
-            if (signerName == null || !signerName.ToLower().Equals(userName.ToLower()))
+            string userName = null;
```
Comment (dubeme) [inline app/wama.web/Controllers/CheckInController.cs]: Nice, but not quite. If this is null, then (SEE NEXT COMMENT)

File: app/wama.web/Controllers/CheckInController.cs [hunk 7]
```diff
@@ -87,10 +87,15 @@ public IActionResult WaiverNotSigned()
...
+            if (userAccount != null)
             {
+                userName = userAccount.FirstName + " " + userAccount.LastName;
+            }
+            
+
+            if (signerName == null || (userName != null && !signerName.ToLower().Equals(userName.ToLower())))
```
Comment (dubeme) [inline app/wama.web/Controllers/CheckInController.cs]: userName.Lower(), will raise a NullReferenceException

**Review by dubeme — CHANGES_REQUESTED** [review]
Still missing the ViewModels tha's used in generating the forms.

Also, still missing some null checks (or perhaps, a better default value).

File: app/wama.web/Controllers/CheckInController.cs (hunk 7 above)
Comment (quazi-irfan) [reply app/wama.web/Controllers/CheckInController.cs]: It won't because && is a short circuit operator.

Comment (dubeme) [reply app/wama.web/Controllers/CheckInController.cs]: Ahh! didn't see that.

Comment (dubeme) [reply app/wama.web/Controllers/CheckInController.cs]: Then just fix the view models, and you're good to go

File: app/wama.web/Views/CheckIn/Index.cshtml (hunk 5 above)
Comment (quazi-irfan) [reply app/wama.web/Views/CheckIn/Index.cshtml:20]: Updated UserAccountViewModel.cs with 4713ba3a33644fa0af881433338bd19c876a15ab

Comment (quazi-irfan) [reply app/wama.web/Views/CheckIn/Index.cshtml:20]: Updated CheckInController.cs with a2f846bee85c740d2db37321a1f5b9318f165e6a

File: app/wama.web/Controllers/CheckInController.cs [hunk 8]
```diff
@@ -29,10 +31,11 @@ public IActionResult Index()
         [HttpPost]
         public async Task<IActionResult> Index(string memberId)
         {
-            if (ModelState.IsValid && !string.IsNullOrWhiteSpace(memberId))
+            if (ModelState.IsValid &&  !string.IsNullOrWhiteSpace(memberId) && (memberId.Length == 6 || memberId.Length == 7))
```
Comment (dubeme) [inline app/wama.web/Controllers/CheckInController.cs]: Nice, but not quite. I think there's a better way to achieve this with attributes

File: app/wama.web/Controllers/CheckInController.cs [hunk 9]
```diff
@@ -76,10 +87,15 @@ public IActionResult WaiverNotSigned()
...
-
-            if (signerName == null || !signerName.Equals(userName))
+            string userName = null;
+            if (userAccount != null)
             {
+                userName = userAccount.FirstName + " " + userAccount.LastName;
+            }
+            
+
+            if (signerName == null || (userName != null && !signerName.ToLower().Equals(userName.ToLower())))
```
Comment (dubeme) [inline app/wama.web/Controllers/CheckInController.cs]: This looks good, but what about something like this.

(string.IsNullOrWhiteSpace(signerName) || string.IsNullOrWhiteSpace(userName) || !signerName.Equals(userName, StringComparison.OrdinalIgnoreCase))

**Review by dubeme — CHANGES_REQUESTED** [review]
See comments

File: app/wama.core/ViewModel/User/UserAccountViewModel.cs [hunk 10]
```diff
@@ -16,7 +16,8 @@ public abstract class UserAccountViewModel
         /// <summary>
         /// Gets or sets the member identifier of this UserAccountViewModel.
         /// </summary>
-        [Display(Name = "ID (7 digit SDSU ID)")]
+        [Display(Name = "ID (6 or 7 digit SDSU ID)")]
+        [MaxLength(7, ErrorMessage = "Member ID length should be either 6 or 7 digits long."), MinLength(6)]
```
Comment (dubeme) [inline app/wama.core/ViewModel/User/UserAccountViewModel.cs]: To make things easier to read, move the MinLength attribute to it's own line

File: app/wama.web/Views/Patron/Create.cshtml [hunk 11]
```diff
@@ -13,8 +13,6 @@
             <label asp-for="MemberId"> </label>
             <input asp-for="MemberId"
                    class="form-control"
-                   pattern="[0-9]{7}"
-                   title="Member ID must be in number and 7 digit long."
```
Comment (dubeme) [inline app/wama.web/Views/Patron/Create.cshtml:17]: Looks like you forgot to put this one back.

**Review by dubeme — CHANGES_REQUESTED** [review]

File: app/wama.web/Views/Patron/Create.cshtml (hunk 11 above)
Comment (quazi-irfan) [reply app/wama.web/Views/Patron/Create.cshtml:17]: If it put it back then the error message in the attribute does not kick in?

Comment (dubeme) [reply app/wama.web/Views/Patron/Create.cshtml:17]: Yep, that's true.

But you're still not seeing/understanding how these validations work.
Or, why they're needed.
Think of, and answer these questions.

* Where should the validation happen if I use a browser that doesn't support HTML5?
* Where should the validation happen if I use a browser that supports HTML5?
* Should the request always have to go back to the server in other for validation to happen?

**Review by dubeme — CHANGES_REQUESTED** [review]

**Comment by sakurano92** [comment]
well quazi, i thought we talked about this last saturday, you put data check method in both view (HTML) and viewmodel (enumerable) side since, 
view gives user why they shouldnt doing it
viemodel gives server and maintenance personal that why there is some strange data try to implemented into.

since html gives the feedback to User, html checking method should remain inthere

and also viewmodel checking is required (which ive seen you implemented) because to make sure not to have some garbage data (or unwanted data) into database when somebody (maintenance, administer so on) try to change something or maintain into it.

**Review by dubeme — APPROVED** [review]

**Pull request merged**